include/srf/segment/resource.hpp

```diff
@@ -0,0 +1,46 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2021-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <srf/segment/object.hpp>
+
+#include <glog/logging.h>
+
+#include <memory>
+#include <ostream>
+#include <utility>
+
+namespace srf::segment {
+
+template <typename ResourceT>
+class Resource final : public Object<ResourceT>
+{
+  public:
+    Resource(std::unique_ptr<ResourceT> resource) : m_resource(std::move(resource)) {}
+    ~Resource() final = default;
+
+  private:
+    ResourceT* get_object() const final
+    {
+        CHECK(m_resource);
+        return m_resource.get();
+    }
+    std::unique_ptr<ResourceT> m_resource;
+};
```

Comment:
What is a `Resource`? Why is it needed?

Reply:
For lack of a better term, a Resource is not a Runnable. 

segment::Object has pure virtual methods. 

Runnable is the specialization of a segment Object that requires a Runner/Launcher. 

A Resource is constructed, owned and the life cycle is managed by the segment.

Reply:
Ok, makes sense. But can we choose a different name? `Resource` is such a highly loaded term.

Reply:
`Component`? `Entity`?

Reply:
`Component` works.

Reply:
done